rpmio/rpmpgp.c

```diff
@@ -503,6 +500,9 @@ static int pgpPrtSubType(const uint8_t *h, size_t hlen, pgpSigType sigtype,
 	case PGPSUBTYPE_REVOKE_REASON:
 	case PGPSUBTYPE_FEATURES:
 	case PGPSUBTYPE_EMBEDDED_SIG:
+	    pgpPrtHex("", p+1, plen-1);
+	    break;
+	case PGPSUBTYPE_NOTATION:
```

Comment:
This was marked resolved but I'm not sure it is. 

What I'm trying to say here is that the only safe assumption wrt these tags falling straight through to default case is to consider them *not recognized*. I think with the semantics explained above, everything downwards of (and including) PGPSUBTYPE_EXPORTABLE_CERT should be flagged not recognized, not just notations.

Reply:
I can change that if you like, but know that RPM will not be able to verify signatures made by a key created by Sequoia PGP then.

Reply:
So Sequoia flags one/some of these packets as critical by default? Which ones?

Reply:
Like I said, I think the only safe assumption about the tags in question is that rpm doesn't have the slightest idea, ie are unrecognized. But from that starting point we can make that conscious decision *here and now*, on record, to ignore the tags that need be, ie kind of the opposite of what happens here.

Reply:
> So Sequoia flags one/some of these packets as critical by default?

Yes.

> Which ones?

- signature creation time
- signature expiration time
- exportable certification
- trust signature
- regular expression
- revocable
- key expiration time
- revocation key
- primary userid
- key flags
- signature target
- embedded signature

Glancing at that list, I'd say relevant for RPM's use case are:

- signature creation time
- signature expiration time
- key expiration time
- revocation key
- key flags
- embedded signature

Reply:
After sleeping over it a bit, what probably bugs me here is this:
If Sequoia marks all those tags as critical then AFAICS rpm is right per the spec to fail on them, yet this is "fixing" something on rather arbitrary grounds. As per the above comments, the actual agenda here is to tweak rpm's interpretation to support Sequoia created keys. State *that* as the reason for the change, and it's totally okay by me.

Reply:
RPM must implement the following:

- Signature creation time
- Signature expiration time
- Key expiration time
- Key flags

In these cases, ignoring that a subpacket is marked critical would be a serious error, as it could cause RPM to accept a package that it must not.  In all four of these cases, the correct fix is for RPM to actually implement the corresponding subpacket, not to ignore it.

RPM can safely ignore the following, even if marked critical, at least for now:

- exportable certification
- trust signature
- regular expression
- revocable
- revocation key
- primary userid
- embedded signature
- preferred symmetric key
- preferred hash algorithm
- preferred compression algorithm
- key server preferences
- policy URL

For notation subpackets, the default must be to reject unknown critical notations, but some notations may be ignorable.  If RPM gets support for revocation or a web of trust, the list of safely ignoreable subpackets will shrink.  Note that unlike subkey binding signatures, *primary* key binding signatures do *not* need to be supported *yet*.  My understanding is that primary key binding signatures ensure Eve cannot claim to have made a signature actually made by Alice, which is not an issue for RPM at this time.

Reply:
> In these cases, ignoring that a subpacket is marked critical would be a serious error, as it could cause RPM to accept a package that it must not. In all four of these cases, the correct fix is for RPM to actually implement the corresponding subpacket, not to ignore it.

I'm not disagreeing, but note that this is already a serious error: RPM ignores these subpackets whether they are marked as critical or not.  Sequoia marking these subpackts as critical merely highlights the problem.

Reply:
> > In these cases, ignoring that a subpacket is marked critical would be a serious error, as it could cause RPM to accept a package that it must not. In all four of these cases, the correct fix is for RPM to actually implement the corresponding subpacket, not to ignore it.
> 
> I'm not disagreeing, but note that this is already a serious error: RPM ignores these subpackets whether they are marked as critical or not. Sequoia marking these subpackts as critical merely highlights the problem.

Would you be willing to fix this?

Reply:
Fix what?

Reply:
Would you be willing to add support to RPM for the four types of subpackets I mentioned above?

Reply:
The RPM developers made it abundantly clear that PGP certificate canonicalization is not needed for RPM.  They also explained to me how OpenPGP works, and how the IETF working group works, and that me spending hours every week working on the revision of the spec is nothing new relative to what Werner did.  All of that made it a lot less exciting to work on RPM.

Also, when I started looking into RPM's implementation, I thought it would be salvageable.  I'm no longer convinced.

Then, I started refactoring RPM with the idea of making the PGP implementation plugable.  That turned out to be very messy and immediately break RPM's public interfaces, therefore it cannot be an incremental improvement.  Not fun.  Lingers in my branch waiting to be picked up again.  May happen someday.

Sorry for dumping.  I'm not going to fix RPM's PGP implementation.

Reply:
Before a OpenPGP certificate can be used, it has to be canonicalized.  Failure to do so properly leads to catastrophic failure of the protocol.  It is at the heart of every PGP implementation.  In Sequoia, this is roughly a fifth of the core library.

Among other things, you need to check that the signatures binding the cert's components to the primary key are valid.  Then, you need a way to reason about the certificate, e.g. given a policy and a time, which (sub)key is eligible to make signatures?

Reply:
I see.  What policy do you suggest for RPM?

Reply:
At the bare minimum, reject unsafe algorithms and key sizes.  But, "It is not RPM's place to impose such a policy".

For inspiration, please see https://docs.rs/sequoia-openpgp/latest/sequoia_openpgp/policy/struct.StandardPolicy.html

Reply:
> The RPM developers made it abundantly clear that PGP certificate canonicalization is not needed for RPM. They also explained to me how OpenPGP works, and how the IETF working group works, and that me spending hours every week working on the revision of the spec is nothing new relative to what Werner did. All of that made it a lot less exciting to work on RPM.

Apologies from my behalf if that's the impression you got. That's not at all what I think, just that we're looking at the thing from very different perspectives: the rpm implementation strives to do the bare minimum to get by, and when you look at a spec through such glasses, I guess things will look very different compared to somebody working on a full implementation.

> Also, when I started looking into RPM's implementation, I thought it would be salvageable. I'm no longer convinced.
> 
> Then, I started refactoring RPM with the idea of making the PGP implementation plugable. That turned out to be very messy and immediately break RPM's public interfaces, therefore it cannot be an incremental improvement. Not fun. Lingers in my branch waiting to be picked up again. May happen someday.
> 
> Sorry for dumping. I'm not going to fix RPM's PGP implementation.

Quite understandable. Thanks for looking into it and the patches provided!